File: shell/gnome/src/refreshTracker.js
export const MIN_SPIN_MS = 800;
export const ERROR_MS = 1000;

export class RefreshTracker {
    constructor() {
        this._inFlight = false;
        this._holdUntil = 0;
        this._errorUntil = 0;
        this._daemonBusy = false;
    }

    press(now) {
        if (this._inFlight || now < this._holdUntil) return false;
        this._inFlight = true;
        this._holdUntil = now + MIN_SPIN_MS;
        this._errorUntil = 0;
        return true;
    }

    settle(succeeded, now) {
        this._inFlight = false;
        if (succeeded) return;
        this._holdUntil = 0;
        this._errorUntil = now + ERROR_MS;
    }

    setDaemonBusy(busy) {
        this._daemonBusy = busy;
    }

    mode(now) {
        if (now < this._errorUntil) return 'error';
        if (this._inFlight || now < this._holdUntil || this._daemonBusy) return 'busy';
        return 'idle';
    }

    nextChange(now) {
        const upcoming = [this._holdUntil, this._errorUntil].filter(deadline => deadline > now);
        return upcoming.length > 0 ? Math.min(...upcoming) : null;
    }
}

File: shell/macos/Sources/HeadroomKit/Presentation/RefreshTracker.swift
import Foundation

public enum RefreshMode: Sendable, Hashable {
    case idle, busy, failed
}

public struct RefreshTracker: Sendable, Hashable {
    public static let minimumSpin: TimeInterval = 0.8
    public static let errorDuration: TimeInterval = 1

    private var inFlight = false
    private var holdUntil = Date.distantPast
    private var errorUntil = Date.distantPast
    private var daemonBusy = false

    public init() {}

    public mutating func press(at now: Date) -> Bool {
        guard !inFlight, now >= holdUntil else { return false }
        inFlight = true
        holdUntil = now.addingTimeInterval(Self.minimumSpin)
        errorUntil = .distantPast
        return true
    }

    public mutating func settle(succeeded: Bool, at now: Date) {
        inFlight = false
        guard !succeeded else { return }
        holdUntil = .distantPast
        errorUntil = now.addingTimeInterval(Self.errorDuration)
    }

    public mutating func setDaemonBusy(_ busy: Bool) {
        daemonBusy = busy
    }

    public func mode(at now: Date) -> RefreshMode {
        if now < errorUntil { return .failed }
        if inFlight || now < holdUntil || daemonBusy { return .busy }
        return .idle
    }

    public func nextChange(after now: Date) -> Date? {
        [holdUntil, errorUntil].filter { $0 > now }.min()
    }
}
